File: apps/kira-studio/internal/postman/url.go
// url.go is P4 D8's half of the boundary: this app's model is URL-authoritative (the raw string
// is canonical), Postman's `url` is a broken-down object. A Go sibling of
// frontend/src/views/httprequest/url.ts, deliberately NOT net/url.Parse — like the TypeScript
// half it must handle a half-formed or {{variable}}-bearing URL without throwing, and net/url has
// its own escaping opinions the round trip must not inherit. The shared rule both files state:
// split on the first '?' and the first '#'.
package postman

import (
	"encoding/json"
	"strings"
)

// SplitURL is the exact counterpart of url.ts's splitUrl.
type SplitURL struct {
	Base string
	// Query is without the leading '?'.
	Query string
	// Hash is without the leading '#'.
	Hash string
}

// Split mirrors url.ts's splitUrl: two IndexOfs, not a parser, so 'api.exa' still splits cleanly.
func Split(text string) SplitURL {
	beforeHash, hash := text, ""
	if i := strings.Index(text, "#"); i >= 0 {
		beforeHash, hash = text[:i], text[i+1:]
	}
	base, query := beforeHash, ""
	if i := strings.Index(beforeHash, "?"); i >= 0 {
		base, query = beforeHash[:i], beforeHash[i+1:]
	}
	return SplitURL{Base: base, Query: query, Hash: hash}
}

// queryPair is one `?a=b` pair, split on the first '='.
type queryPair struct{ Key, Value string }

func parseQuery(query string) []queryPair {
	if query == "" {
		return nil
	}
	out := []queryPair{}
	for _, pair := range strings.Split(query, "&") {
		if pair == "" {
			continue
		}
		if i := strings.Index(pair, "="); i >= 0 {
			out = append(out, queryPair{Key: pair[:i], Value: pair[i+1:]})
			continue
		}
		out = append(out, queryPair{Key: pair})
	}
	return out
}

// hasScheme mirrors url.ts's withScheme test without adding one — this side never invents a
// protocol the user did not type.
func hasScheme(base string) bool {
	i := strings.Index(base, "://")
	if i <= 0 {
		return false
	}
	for j, r := range base[:i] {
		switch {
		case r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z':
		case j > 0 && (r >= '0' && r <= '9' || r == '+' || r == '.' || r == '-'):
		default:
			return false
		}
	}
	return true
}

// containsVariable reports whether s carries a {{name}} reference — P5's own syntax, which must
// survive Build's splitting untouched (D8).
func containsVariable(s string) bool { return strings.Contains(s, "{{") }

// Build turns a raw URL string into Postman's own broken-down `url` object — what Postman itself
// writes in v2.1, and the shape with no ambiguity about how it is consumed. Empty members are
// omitted; `raw` is always present, which is what makes Import(Build(s)) == s by construction.
func Build(raw string) map[string]json.RawMessage {
	out := map[string]json.RawMessage{"raw": mustRaw(raw)}
	parts := Split(raw)

	base := parts.Base
	if hasScheme(base) {
		i := strings.Index(base, "://")
		out["protocol"] = mustRaw(base[:i])
		base = base[i+3:]
	}

	hostPort, path := base, ""
	if i := strings.Index(base, "/"); i >= 0 {
		hostPort, path = base[:i], base[i+1:]
	}

	if hostPort != "" {
		host := hostPort
		// A trailing ":<digits>" is a port; ':' inside a {{variable}} is not.
		if i := strings.LastIndex(hostPort, ":"); i >= 0 && !containsVariable(hostPort[i:]) {
			if port := hostPort[i+1:]; port != "" && isAllDigits(port) {
				host = hostPort[:i]
				out["port"] = mustRaw(port)
			}
		}
		if host != "" {
			out["host"] = mustRaw(hostSegments(host))
		}
	}

	if path != "" {
		out["path"] = mustRaw(strings.Split(path, "/"))
	}

	if pairs := parseQuery(parts.Query); len(pairs) > 0 {
		entries := make([]map[string]string, 0, len(pairs))
		for _, p := range pairs {
			entries = append(entries, map[string]string{"key": p.Key, "value": p.Value})
		}
		out["query"] = mustRaw(entries)
	}

	if parts.Hash != "" {
		out["hash"] = mustRaw(parts.Hash)
	}
	return out
}

// hostSegments splits a host on '.', except when it carries a {{variable}} — `{{baseUrl}}` is one
// segment, not three (D8).
func hostSegments(host string) []string {
	if containsVariable(host) {
		return []string{host}
	}
	return strings.Split(host, ".")
}

func isAllDigits(s string) bool {
	for _, r := range s {
		if r < '0' || r > '9' {
			return false
		}
	}
	return len(s) > 0
}

// ImportURL turns Postman's `url` member — a string or an object (F2) — back into this app's raw
// URL string. An object's own `raw` wins when it is present and non-empty, which is the common
// case and which preserves `:pathVariable` segments and `{{baseUrl}}` references exactly as typed.
func ImportURL(raw json.RawMessage) string {
	if len(raw) == 0 {
		return ""
	}
	if s, ok := decodeString(raw); ok {
		return s
	}
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(raw, &obj); err != nil {
		return ""
	}
	if s, ok := decodeString(obj["raw"]); ok && s != "" {
		return s
	}
	return reconstructURL(obj)
}

// reconstructURL rebuilds the string from protocol/host/port/path/query/hash, with `host` and
// `path` each handling both their string and array forms (F2). A `query` entry with
// `disabled: true` is omitted — a disabled param has no representation in a URL-authoritative
// model — and survives in origin regardless.
func reconstructURL(obj map[string]json.RawMessage) string {
	var b strings.Builder
	if protocol, ok := decodeString(obj["protocol"]); ok && protocol != "" {
		b.WriteString(protocol)
		b.WriteString("://")
	}
	if host := joinStringOrArray(obj["host"], "."); host != "" {
		b.WriteString(host)
	}
	if port, ok := decodeScalarString(obj["port"]); ok && port != "" {
		b.WriteString(":")
		b.WriteString(port)
	}
	if path := joinStringOrArray(obj["path"], "/"); path != "" {
		if !strings.HasPrefix(path, "/") {
			b.WriteString("/")
		}
		b.WriteString(path)
	}
	if query := reconstructQuery(obj["query"]); query != "" {
		b.WriteString("?")
		b.WriteString(query)
	}
	if hash, ok := decodeString(obj["hash"]); ok && hash != "" {
		b.WriteString("#")
		b.WriteString(hash)
	}
	return b.String()
}

func reconstructQuery(raw json.RawMessage) string {
	if len(raw) == 0 {
		return ""
	}
	var entries []map[string]json.RawMessage
	if err := json.Unmarshal(raw, &entries); err != nil {
		return ""
	}
	parts := make([]string, 0, len(entries))
	for _, e := range entries {
		if disabled, ok := decodeBool(e["disabled"]); ok && disabled {
			continue
		}
		key, _ := decodeScalarString(e["key"])
		value, hasValue := decodeScalarString(e["value"])
		if !hasValue {
			parts = append(parts, key)
			continue
		}
		parts = append(parts, key+"="+value)
	}
	return strings.Join(parts, "&")
}

// joinStringOrArray handles F2's `host`/`path` oneOf: a bare string, or an array whose entries are
// strings or path-variable segment objects ({"type":"string","value":":id"}), which contribute
// their `value`.
func joinStringOrArray(raw json.RawMessage, sep string) string {
	if len(raw) == 0 {
		return ""
	}
	if s, ok := decodeString(raw); ok {
		return s
	}
	var entries []json.RawMessage
	if err := json.Unmarshal(raw, &entries); err != nil {
		return ""
	}
	parts := make([]string, 0, len(entries))
	for _, e := range entries {
		if s, ok := decodeString(e); ok {
			parts = append(parts, s)
			continue
		}
		var seg map[string]json.RawMessage
		if err := json.Unmarshal(e, &seg); err != nil {
			continue
		}
		if v, ok := decodeScalarString(seg["value"]); ok {
			parts = append(parts, v)
		}
	}
	return strings.Join(parts, sep)
}

File: packages/api-core/src/http/url.ts
// D9: pure, DOM-free URL helpers — no URLSearchParams. URLSearchParams.toString() encodes a
// space as '+', so any round trip through the Params table would silently rewrite a user's %20;
// these hand-written functions never do. The rule this file exists to make testable: typing in
// the URL field updates the Params table and never rewrites the URL; editing the table rewrites
// the URL (HttpRequestView.vue owns that direction — this module only splits/builds strings).

import { splitTemplateSpans } from './substitute';

export interface SplitUrl {
  base: string;
  /** Without the leading '?'. */
  query: string;
  /** Without the leading '#'. */
  hash: string;
}

/** Two indexOfs, not new URL() — a half-typed URL ('api.exa') must still split cleanly. */
export function splitUrl(text: string): SplitUrl {
  const hashIdx = text.indexOf('#');
  const beforeHash = hashIdx >= 0 ? text.slice(0, hashIdx) : text;
  const hash = hashIdx >= 0 ? text.slice(hashIdx + 1) : '';
  const queryIdx = beforeHash.indexOf('?');
  const base = queryIdx >= 0 ? beforeHash.slice(0, queryIdx) : beforeHash;
  const query = queryIdx >= 0 ? beforeHash.slice(queryIdx + 1) : '';
  return { base, query, hash };
}

export interface QueryPair {
  name: string;
  value: string;
  /** P21 round 3 functional finding 11: true when the pair's own text had no `=` at all (a bare
   *  `?flag`, not `?flag=`). buildQuery uses this to rebuild a bare flag as `flag` rather than
   *  always appending `=` — without it, any *other* row changing in the same table rewrote the
   *  whole query string (HttpRequestView.vue does this on every edit) and silently turned a
   *  valueless flag the user never touched into `flag=`, a different request for any server that
   *  distinguishes "no value" from "empty value". Optional: a freshly added row (blankParam()) has
   *  no occasion to be bare, and undefined behaves exactly like false. Once `value` becomes
   *  non-empty (the user typed one), the pair is no longer bare regardless of this flag — see
   *  buildQuery below. */
  bare?: boolean;
}

// P21 round 3 functional finding 11: `+` used to be asymmetric — left untouched here on decode
// (deliberately, per this file's own header: a space must round-trip as %20, never silently become
// `+`), but encoded to `%2B` by encodeQueryComponent below (plain encodeURIComponent's own
// behaviour). Any *other* row changing in the same Params table rewrites the whole query string
// (HttpRequestView.vue does this on every edit), so a literal `+` a user typed — `q=hello+world`,
// the overwhelming majority of form-encoded readers' own space encoding — silently became
// `q=hello%2Bworld` the moment an unrelated param changed, a different request for any server that
// treats `+` as a space. Resolved by leaving `+` untouched on *both* sides (this decoder already
// did; encodeQueryComponent now un-encodes the one sequence encodeURIComponent can produce for it)
// rather than adopting `+`-means-space, which is exactly what this file's header comment already
// rejected for the more common %20 case.
function decodeComponent(s: string): string {
  try {
    return decodeURIComponent(s);
  } catch {
    // A malformed escape (a bare '%' from mid-typing) passes through raw rather than throwing.
    return s;
  }
}

/** Split on '&', then the first '=' of each pair — `parseQuery('a=1&b') → [{a,1},{b,''}]`. */
export function parseQuery(query: string): QueryPair[] {
  if (!query) return [];
  return query
    .split('&')
    .filter((pair) => pair.length > 0)
    .map((pair) => {
      const eq = pair.indexOf('=');
      const rawName = eq >= 0 ? pair.slice(0, eq) : pair;
      const rawValue = eq >= 0 ? pair.slice(eq + 1) : '';
      return { name: decodeComponent(rawName), value: decodeComponent(rawValue), bare: eq < 0 };
    });
}

/** encodeURIComponent each half — never URLSearchParams.toString(), which encodes a space as
 *  '+' instead of '%20' and would silently rewrite what the user typed. Finding 16: a `{{name}}`
 *  reference living inside a name or value is left untouched rather than being encoded into
 *  `%7B%7Bname%7D%7D`, a form neither substitution engine (this package's own `resolve`, nor
 *  internal/apivars/resolve.go) recognises any more. Finding 11 (round 3): `%2B` is un-encoded
 *  back to a literal `+` afterward — the only percent sequence encodeURIComponent can produce for
 *  it, so this is exact and never touches a `+` produced by encoding some other original byte. */
function encodeQueryComponent(s: string): string {
  return splitTemplateSpans(s)
    .map((span) =>
      span.isReference ? span.text : encodeURIComponent(span.text).replace(/%2B/g, '+'),
    )
    .join('');
}

export function buildQuery(pairs: readonly QueryPair[]): string {
  return pairs
    .map((p) =>
      // Finding 11 (round 3): a bare `?flag` (no `=` at all) must rebuild bare — always appending
      // `=` turned a valueless flag into `flag=` the moment any other row's edit rewrote the query
      // string, a different request for a server that distinguishes the two. Once the user has
      // actually typed a value, the pair is no longer bare regardless of how it started.
      p.bare && p.value === ''
        ? encodeQueryComponent(p.name)
        : `${encodeQueryComponent(p.name)}=${encodeQueryComponent(p.value)}`,
    )
    .join('&');
}

function withScheme(base: string): string {
  return /^[a-zA-Z][a-zA-Z0-9+.-]*:\/\//.test(base) ? base : `https://${base}`;
}

/** D2: the URL's path, else its host, else the raw text, else 'New request' — TabKindDef.title
 *  and the view header both call this. P4 D14: a saved request's own name wins over all of it, so
 *  "Create order" stops rendering as /v2/orders — one line here keeps the function pure and keeps
 *  both consumers unchanged, rather than teaching either about collections. */
export function httpRequestTitle(state: { url: string; name?: string }): string {
  if (state.name) return state.name;
  const url = state.url.trim();
  if (!url) return 'New request';
  const { base } = splitUrl(url);
  try {
    const parsed = new URL(withScheme(base));
    if (parsed.pathname.length > 1) return parsed.pathname;
    if (parsed.host) return parsed.host;
  } catch {
    // Not parseable yet (mid-typing) — fall through to the raw text.
  }
  return url;
}
